### DISCUSSION TITLE
Difference Between Git Commands vs GitHub Desktop to Review Pull Requests

### DISCUSSION THREAD

**Discussion by JessicaLucindaCheng** [hackforla/website#2618] [General]
Note: I am deleting branch update-project-profile-design-systems-2419 associated with pull request 2556 because it has already been merged into the repo. Thus, it won't be possible to test/try out the examples below.

[Pull request 2556 Updated design systems' project profile with current project team](https://github.com/hackforla/website/pull/2556) as will be used as an example. An example of a page that is different when using git commands vs GitHub Desktop is localhost:4000/projects/design-systems. You'll notice the links, languages, and technologies look different. Also, I am assuming I have fetched and merged the most recent updates from the Hack for LA repo into my forked gh-pages repo before doing any of the following.

**Git Commands**

- When I use git commands, the following happens:

  - `git checkout -b JessicaLucindaCheng-update-project-profile-design-systems-2419 gh-pages`
   A new branch named "JessicaLucindaCheng-update-project-profile-design-systems-2419" is created based on my forked gh-pages, which is up to date with the Hack for LA repo. That command also switched into the new branch JessicaLucindaCheng-update-project-profile-design-systems-2419 that was created. [1]

  - `git pull https://github.com/JessicaLucindaCheng/website.git update-project-profile-design-systems-2419`
  This fetches content from the update-project-profile-design-systems-2419 branch in JessicaLucindaCheng's forked repo and merges that content into my new branch JessicaLucindaCheng-update-project-profile-design-systems-2419. [2, 3]

  - Since the JessicaLucindaCheng-update-project-profile-design-systems-2419 was based on my branch (gh-pages) with the most recent Hack for LA repo updates, the changes JessicaLucindaCheng made along with the most recent updates are seen together. This occurs even if JessicaLucindaCheng did not incorporate the most recent changes into her branch.

**GitHub Desktop**

- Using GitHub Desktop to review pull requests is similar to using the following git commands

  - `git fetch upstream pull/2556/head:mytestingpr`
  where upstream is the gh-pages from the Hack for LA repo, 2556 is the pull request number, and "mytestingpr" is the name of a new branch being created. Based on the PR number 2556, this line of code fetches the contents of the branch update-project-profile-design-systems-2419 from JessicaLucindaChengs's forked repo and merges it into the empty mytestingpr branch.  [4]
  - `git checkout mytestingpr`
  This switches into the mytestingpr branch. The contents of the mytestingpr branch are exactly same as the one in JessicaLucindaCheng's forked repo branch update-project-profile-design-systems-2419. This may not contain the most recent updates depending on if JessicaLucindaCheng included the updates in her branch and depending on when I am reviewing/pulling the pull request, such as if it has been a few days since the pull request was created. [4]

References
[1] New Branches section of https://www.atlassian.com/git/tutorials/using-branches/git-checkout
[2] Git Pull on Remotes section of https://www.atlassian.com/git/tutorials/syncing/git-pull
[3] https://stackoverflow.com/questions/1709177/git-pull-a-certain-branch-from-github
[4] Steps 5 and 6 of https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally

**Comment by SAUMILDHANKAR** [comment]
Thanks Jessica for starting this discussion. Is there a command in GitHub Desktop to make sure recent updates are incorporated as well (refering to the last line in your post)?
Also, do you think it would be helpful if we add the instructions for GitHub Desktop on our wiki page: https://github.com/hackforla/website/wiki/How-to-Review-Pull-Requests? Since, not many currently in our team are using GitHub Desktop, I think you would be the best judge for it if it might help someone in future who is using the GitHub desktop.

**Reply by JessicaLucindaCheng** [reply]
- Just to clarify what I wrote previously:
  - I was assuming I fetched and merged the most recent updates from the Hack for LA repo into my forked gh-pages repo before doing any of the commands that I wrote about.
  - In the last line of my post, I made a correction saying it also depends on when I am reviewing/pulling the pull request, such as if it has been a few days since the pull request was created.


- I don't know how to check if the most recent updates were incorporated using git commands. If I was using GitHub Desktop, I would use the "Compare to branch" feature to compare the pull request branch to my upstream-gh-pages branch, which is my most up to date branch. This will show how many commits behind the pr branch is compared to the upstream-gh-pages and how many commits ahead. If there are no commits behind, then the pr branch has the most recent updates/changes. However, this does not necessarily indicate if the pr creator incorporated the most recent updates from the Hack for LA repo at the time of pr creation. For example, if I created a pr on Dec 7 at 11am and incorporated the most recent changes at that time, a pr reviewer viewing it today would see my pr is behind and does not include any updates that occurred after Dec 7 at 11am.
- If I'm just checking if recent updates were incorporated at time of pr creation, after using the "Compare to branch" feature, I just look at how many days ago the pr was created and then look at the how many days ago the oldest "behind" commit is. So, if a pr was created 6 days ago and the oldest "behind" commit is less than 6 days, it probably means the most recent updates at the time of pr creation were incorporated.


- I don't think it would be helpful incorporating these instructions for GitHub Desktop into the wiki because:
1) The whole wiki uses git commands and talking about GitHub Desktop in just one section would be confusing.
2) Most people don't use GitHub Desktop.
3) I think the current git commands that we use are the better option. My reason is as long as my forked gh-pages are up to date and I'm using the current git commands for reviewing pull requests, I will have the most recent updates incorporated into the pull request I'm reviewing. Then, I can see how the updates and pr changes impact each other.
